zetaclient/bitcoin/fee.go

```diff
@@ -0,0 +1,234 @@
+	// number of transaction inputs and outputs.
+	// #nosec G701 always positive
+	return uint64(8 + wire.VarIntSerializeSize(numInputs) + wire.VarIntSerializeSize(numOutputs))
+}
+
+// EstimateOuttxSize estimates the size of a outtx in vBytes
+func EstimateOuttxSize(numInputs uint64, payees []btcutil.Address) uint64 {
+	if numInputs == 0 {
+		return 0
+	}
+	// #nosec G701 always positive
+	numOutputs := 2 + uint64(len(payees))
+	bytesWiredTx := WiredTxSize(numInputs, numOutputs)
+	bytesInput := numInputs * bytesPerInput
+	bytesOutput := uint64(2) * bytesPerOutputP2WPKH // new nonce mark, change
+
+	// calculate the size of the outputs to payees
+	bytesToPayees := uint64(0)
+	for _, to := range payees {
+		bytesToPayees += GetOutputSizeByAddress(to)
+	}
+	// calculate the size of the witness
+	bytesWitness := bytes1stWitness + (numInputs-1)*bytesPerWitness
+	// https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#transaction-size-calculations
+	// Calculation for signed SegWit tx: blockchain.GetTransactionWeight(tx) / 4
+	return bytesWiredTx + bytesInput + bytesOutput + bytesToPayees + bytesWitness/blockchain.WitnessScaleFactor
+}
+
+// GetOutputSizeByAddress returns the size of a tx output in bytes by the given address
+func GetOutputSizeByAddress(to btcutil.Address) uint64 {
+	switch addr := to.(type) {
+	case *bitcoin.AddressTaproot:
+		if addr == nil {
+			return 0
+		}
+		return bytesPerOutputP2TR
+	case *btcutil.AddressWitnessScriptHash:
+		if addr == nil {
+			return 0
+		}
+		return bytesPerOutputP2WSH
+	case *btcutil.AddressWitnessPubKeyHash:
+		if addr == nil {
+			return 0
+		}
+		return bytesPerOutputP2WPKH
+	case *btcutil.AddressScriptHash:
+		if addr == nil {
+			return 0
+		}
+		return bytesPerOutputP2SH
+	case *btcutil.AddressPubKeyHash:
+		if addr == nil {
+			return 0
+		}
+		return bytesPerOutputP2PKH
+	default:
+		return bytesPerOutputP2WPKH
+	}
+}
+
+// OuttxSizeDepositor returns outtx size (68vB) incurred by the depositor
+func OuttxSizeDepositor() uint64 {
+	return bytesPerInput + bytesPerWitness/blockchain.WitnessScaleFactor
+}
+
+// OuttxSizeWithdrawer returns outtx size (177vB) incurred by the withdrawer (1 input, 3 outputs)
+func OuttxSizeWithdrawer() uint64 {
+	bytesWiredTx := WiredTxSize(1, 3)
+	bytesInput := uint64(1) * bytesPerInput         // nonce mark
+	bytesOutput := uint64(2) * bytesPerOutputP2WPKH // 2 P2WPKH outputs: new nonce mark, change
+	bytesOutput += bytesPerOutputAvg                // 1 output to withdrawer's address
+	return bytesWiredTx + bytesInput + bytesOutput + bytes1stWitness/blockchain.WitnessScaleFactor
+}
+
+// DepositorFee calculates the depositor fee in BTC for a given sat/byte fee rate
+// Note: the depositor fee is charged in order to cover the cost of spending the deposited UTXO in the future
+func DepositorFee(satPerByte int64) float64 {
+	return float64(satPerByte) * float64(BtcOutTxBytesDepositor) / btcutil.SatoshiPerBitcoin
```

Comment:
Is it standard in Bitcoin development working with float64? Type is not deterministic so we should avoid doing operation with it when possible

Reply:
Yeah, the standard Bitcoin output value is float64 and the conversion is necessary.